kernel/arch/x86/cpuid.c

```diff
@@ -0,0 +1,329 @@
+	c->extended_cpuid_level = eax;
+
+	if ((eax & 0xffff0000) == CPUID_EXTENDED) {
+		if (eax >= CPUID_FEATURES) {
+			cpuid(CPUID_FEATURES, &eax, &ebx, &ecx, &edx);
+			c->x86_capability[CPUID_8000_0001_ECX] = ecx;
+			c->x86_capability[CPUID_8000_0001_EDX] = edx;
+		}
+	}
+
+	if (c->extended_cpuid_level >= CPUID_ADVANCEDPOWER) {
+		cpuid(CPUID_ADVANCEDPOWER, &eax, &ebx, &ecx, &edx);
+		c->x86_capability[CPUID_8000_0007_EBX] = ebx;
+		c->x86_power = edx;
+	}
+
+	if (c->extended_cpuid_level >= CPUID_VIRTPHYADDR) {
+		cpuid(CPUID_VIRTPHYADDR, &eax, &ebx, &ecx, &edx);
+		c->x86_virt_bits = (eax >> 8) & 0xff;
+		c->x86_phys_bits = eax & 0xff;
+		c->x86_capability[CPUID_8000_0008_EBX] = ebx;
+	}
+
+	if (c->extended_cpuid_level >= 0x8000000a)
+		c->x86_capability[CPUID_8000_000A_EDX] = cpuid_edx(0x8000000a);
+
+}
+
+void cpu_detect_cache_sizes(struct cpuinfo_x86 *c)
+{
+	unsigned int n, dummy, ebx, ecx, edx, l2size;
+
+	n = c->extended_cpuid_level;
+
+	if (n >= 0x80000005) {
+		cpuid(0x80000005, &dummy, &ebx, &ecx, &edx);
+		c->x86_cache_size = (ecx>>24) + (edx>>24);
+
+		/* For 64 bits processor. See cpuid.h. */
+		//c->x86_tlbsize = 0;
+	}
+
+	if (n < 0x80000006)     /* Some chips just has a large L1. */
+		return;
+
+	cpuid(0x80000006, &dummy, &ebx, &ecx, &edx);
+	l2size = ecx >> 16;
+
+	/* For 64 bits processor. See cpuid.h. */
+	//c->x86_tlbsize += ((ebx >> 16) & 0xfff) + (ebx & 0xfff);
+
+	if (l2size == 0)
+		return;         /* Again, no L2 cache is possible */
+
+	c->x86_cache_size = l2size;
+}
+
+void cpu_detect(struct cpuinfo_x86 *c)
+{
+	cpuid(CPUID_GETVENDORSTRING,
+		(unsigned int *)&c->cpuid_level,
+		(unsigned int *)&c->x86_vendor_id[0],
+		(unsigned int *)&c->x86_vendor_id[8],
+		(unsigned int *)&c->x86_vendor_id[4]);
+	
+	c->x86 = 4;
+	if (c->cpuid_level >= CPUID_GETFEATURES) {
+		uint32_t junk, tfms, cap0, misc;
+		
+		cpuid(CPUID_GETFEATURES, &tfms, &misc, &junk, &cap0);
+		c->x86          = x86_family(tfms);
+		c->x86_model    = x86_model(tfms);
+		c->x86_stepping = x86_stepping(tfms);
+		
+		if (cap0 & (1<<19)) {
+			c->x86_clflush_size = ((misc >> 8) & 0xff) * 8;
+			c->x86_cache_alignment = c->x86_clflush_size;
+		}
+	}
+}
+
+static void get_model_name(struct cpuinfo_x86 *c)
+{
+	char *p, *q, *s;
+	uint32_t eax, ebx, ecx, edx;
+	
+	if (c->extended_cpuid_level < CPUID_BRANDSTRINGEND)
+		return;
+	cpuid(CPUID_BRANDSTRING, &eax, &ebx, &ecx, &edx);
+	__regs_to_str(eax, ebx, ecx, edx, &c->x86_model_id[0]);
+	cpuid(CPUID_BRANDSTRINGMORE, &eax, &ebx, &ecx, &edx);
+	__regs_to_str(eax, ebx, ecx, edx, &c->x86_model_id[16]);
+	cpuid(CPUID_BRANDSTRINGEND, &eax, &ebx, &ecx, &edx);
+	__regs_to_str(eax, ebx, ecx, edx, &c->x86_model_id[32]);
+	c->x86_model_id[48] = '\0';
+	
+	/* Trim whitespace */
+	p = q = s = &c->x86_model_id[0];
+	
+	while (*p == ' ')
+		p++;
+	
+	while (*p) {
+		/* Note the last non-whitespace index */
+		if (*p != ' ')
+			s = q;
+		*q++ = *p++;
+	}
+	
+	*(s + 1) = '\0';
+}
+
+void cpuinfo(void)
```

Comment:
Although TempOS supports only x86 so far, I would expect to see some independent architecture "export" from here. For instance, how this information can be accessed from main kernel thread? This PR would be much more complete if you can provide not only the dependent part, but the independent interface for cpu information as well....